File: julia/src/models/bess_supervisor.jl
"""
Rule-based BESS supervisor with adaptive set-point.

States:
  IDLE       — within freq dead-band & SOC mid-range
  DISCHARGE  — Δω<0 (under-frequency) and SOC>SOC_min, supports inertia
  CHARGE     — Δω>0 or P_pv > P_load proxy and SOC<SOC_max
  STANDBY    — SOC at limits, command zero
"""
@enum BessMode IDLE DISCHARGE CHARGE STANDBY

function bess_supervisor(Δω::Float64, dωdt::Float64, P_pv::Float64,
                         P_load::Float64, SOC::Float64, p::SystemParams)
    if SOC <= p.SOC_min + 0.02
        return (mode=STANDBY, P_cmd=0.0)
    elseif SOC >= p.SOC_max - 0.02 && P_pv > P_load
        return (mode=STANDBY, P_cmd=0.0)
    end

    if abs(Δω) < p.Δω_dead && abs(dωdt) < 2pi*0.05
        # not a frequency event
        if P_pv > P_load && SOC < p.SOC_max
            P_cmd = -min(p.P_bess_max, P_pv - P_load)
            return (mode=CHARGE, P_cmd=P_cmd)
        else
            return (mode=IDLE, P_cmd=0.0)
        end
    end

    # frequency event → support
    P_support = -p.P_bess_max * tanh(2*Δω/(2pi*0.5) + 0.5*dωdt/(2pi*0.5))
    if P_support > 0 && SOC > p.SOC_min
        return (mode=DISCHARGE, P_cmd=P_support)
    elseif P_support < 0 && SOC < p.SOC_max
        return (mode=CHARGE, P_cmd=P_support)
    else
        return (mode=STANDBY, P_cmd=0.0)
    end
end

File: matlab/functions/bess_supervisor.m
function [mode, P_bess_cmd] = bess_supervisor(domega, domega_dt, P_pv, P_load, SOC, params)
% BESS_SUPERVISOR  4-mode state machine (IDLE/DISCHARGE/CHARGE/STANDBY)
%   Sumber: julia/src/models/bess_supervisor.jl
%
% Mode encoding: 0=IDLE, 1=DISCHARGE, 2=CHARGE, 3=STANDBY

    domega_dead = 2*pi*0.02;     % 0.02 Hz dead-band
    rocof_dead  = 2*pi*0.05;     % 0.05 Hz/s dead-band

    % --- 1. STANDBY (SOC limit protection) ---
    if (SOC <= params.SOC_min + 0.02) || ...
       (SOC >= params.SOC_max - 0.02 && P_pv > P_load)
        mode = 3;
        P_bess_cmd = 0;
        return;
    end

    % --- 2. Normal operation (no freq event) ---
    if abs(domega) < domega_dead && abs(domega_dt) < rocof_dead
        if P_pv > P_load && SOC < params.SOC_max
            mode = 2;  % CHARGE (absorb surplus)
            P_bess_cmd = -min(params.P_bess_max, P_pv - P_load);
        else
            mode = 0;  % IDLE
            P_bess_cmd = 0;
        end
        return;
    end

    % --- 3. Frequency event response ---
    sev_w = 2 * domega / (2*pi*0.5);
    sev_r = 0.5 * domega_dt / (2*pi*0.5);
    P_support = -params.P_bess_max * tanh(sev_w + sev_r);

    if P_support > 0 && SOC > params.SOC_min
        mode = 1;  % DISCHARGE
        P_bess_cmd = P_support;
    elseif P_support < 0 && SOC < params.SOC_max
        mode = 2;  % CHARGE
        P_bess_cmd = P_support;
    else
        mode = 3;  % STANDBY (tidak bisa support, SOC limit)
        P_bess_cmd = 0;
    end
end
